File: src/lib/BaseGiven.ts
/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable @typescript-eslint/no-explicit-any */

// Do not add logging to this file as it is used by the pure runtime.

import type { Ibdd_in_any, Ibdd_out_any } from "../CoreTypes";

import { ITestArtifactory, ITLog, ITTestResourceConfiguration } from ".";
import { IPM } from "./types.js";
import { afterEachProxy, beforeEachProxy } from "./pmProxy.js";
import { BaseSuite } from "./BaseSuite";

/**
 * Represents a collection of Given conditions keyed by their names.
 * Givens are typically organized as named collections because:
 * - They set up different initial states for tests
 * - Tests often need to reference specific Given conditions by name
 * - This allows for better organization and reuse of setup logic
 * - The BDD pattern often involves multiple named Given scenarios
 */
export type IGivens<I extends Ibdd_in_any> = Record<string, BaseGiven<I>>;

export abstract class BaseGiven<I extends Ibdd_in_any> {
  features: string[];
  whens: any[];
  thens: any[];
  error: Error;
  fail: any;
  store: I["istore"];
  recommendedFsPath: string;
  givenCB: I["given"];
  initialValues: any;
  key: string;
  failed: boolean;
  artifacts: string[] = [];
  fails: number;

  status: boolean | undefined;

  addArtifact(path: string) {
    if (typeof path !== "string") {
      throw new Error(
        `[ARTIFACT ERROR] Expected string, got ${typeof path}: ${JSON.stringify(
          path
        )}`
      );
    }
    const normalizedPath = path.replace(/\\/g, "/"); // Normalize path separators
    this.artifacts.push(normalizedPath);
  }

  constructor(
    features: string[],
    whens: any[],
    thens: any[],
    givenCB: I["given"],
    initialValues: any
  ) {
    this.features = features;
    this.whens = whens;
    this.thens = thens;
    this.givenCB = givenCB;
    this.initialValues = initialValues;
    this.fails = 0; // Initialize fail count
  }

  beforeAll(store: I["istore"]) {
    return store;
  }

  toObj() {
    return {
      key: this.key,
      whens: (this.whens || []).map((w) => {
        if (w && w.toObj) return w.toObj();
        console.error("When step is not as expected!", JSON.stringify(w));
        return {};
      }),
      thens: (this.thens || []).map((t) => (t && t.toObj ? t.toObj() : {})),
      error: this.error ? [this.error, this.error.stack] : null,
      failed: this.failed,
      features: this.features || [],
      artifacts: this.artifacts,
      status: this.status,
    };
  }

  abstract givenThat(
    subject: I["isubject"],
    testResourceConfiguration,
    artifactory: ITestArtifactory,
    givenCB: I["given"],
    initialValues: any,
    pm: IPM
  ): Promise<I["istore"]>;

  async afterEach(
    store: I["istore"],
    key: string,
    artifactory: ITestArtifactory,
    pm: IPM
  ): Promise<I["istore"]> {
    return store;
  }

  abstract uberCatcher(e: any): void;

  async give(
    subject: I["isubject"],
    key: string,
    testResourceConfiguration: ITTestResourceConfiguration,
    tester: (t: Awaited<I["then"]> | undefined) => boolean,
    artifactory: ITestArtifactory,
    tLog: ITLog,
    pm: IPM,
    suiteNdx: number
  ) {
    this.key = key;
    this.fails = 0; // Initialize fail count for this given

    tLog(`\n ${this.key}`);
    tLog(`\n Given: ${this.key}`);

    const givenArtifactory = (fPath: string, value: unknown) =>
      artifactory(`given-${key}/${fPath}`, value);

    this.uberCatcher((e: any) => {
      console.error(e.toString());
      this.error = e.error;
      tLog(e.stack);
    });

    try {
      // Ensure addArtifact is properly bound to 'this'
      const addArtifact = this.addArtifact.bind(this);
      const proxiedPm = beforeEachProxy(pm, suiteNdx.toString(), addArtifact);
      this.store = await this.givenThat(
        subject,
        testResourceConfiguration,
        givenArtifactory,
        this.givenCB,
        this.initialValues,
        proxiedPm
      );
      this.status = true;
    } catch (e: any) {
      this.status = false;

      this.failed = true;
      this.fails++; // Increment fail count
      this.error = e.stack;
      // throw e;
    }

    try {
      const whens = this.whens || [];

      if (whens.length > 0) {
        for (const [whenNdx, whenStep] of whens.entries()) {
          try {
            this.store = await whenStep.test(
              this.store,
              testResourceConfiguration,
              tLog,
              pm,
              `suite-${suiteNdx}/given-${key}/when/${whenNdx}`
            );
          } catch (e) {
            this.failed = true;
            this.fails++;
            throw e;
          }
        }
      }

      for (const [thenNdx, thenStep] of this.thens.entries()) {
        try {
          const t = await thenStep.test(
            this.store,
            testResourceConfiguration,
            tLog,
            pm,
            `suite-${suiteNdx}/given-${key}/then-${thenNdx}`
          );
          tester(t);
        } catch (e: any) {
          this.failed = true;
          this.fails++; // Increment fail count
          // Re-throw to propagate the error
          throw e;
        }
      }
    } catch (e: any) {
      this.error = e.stack;
      this.failed = true;
    } finally {
      try {
        const addArtifact = this.addArtifact.bind(this);
        const proxiedPm = afterEachProxy(
          pm,
          suiteNdx.toString(),
          key,
          addArtifact
        );
        await this.afterEach(this.store, this.key, givenArtifactory, proxiedPm);
      } catch (e: any) {
        this.failed = true;
        this.fails++; // Increment fail count
        throw e;
        // this.error = e.message;
      }
    }

    return this.store;
  }
}

File: pitono/base_given.py
from typing import List, Dict, Any, Callable, Optional
from .types import Isubject, Istore, Iselection, Then, Given

class BaseGiven:
    def __init__(
        self,
        name: str,
        features: List[str],
        whens: List[Any],
        thens: List[Any],
        given_cb: Given,
        initial_values: Any
    ):
        self.name = name
        self.features = features or []
        self.whens = whens or []
        self.thens = thens or []
        self.given_cb = given_cb
        self.initial_values = initial_values
        self.artifacts: List[str] = []
        self.error: Optional[Exception] = None
        self.failed = False
        self.store: Optional[Istore] = None
        self.key: Optional[str] = None
    
    def add_artifact(self, path: str) -> None:
        # Normalize path separators
        normalized_path = path.replace('\\', '/')
        self.artifacts.append(normalized_path)
    
    def to_obj(self) -> Dict[str, Any]:
        return {
            'key': self.key,
            'name': self.name,
            'whens': [w.to_obj() if hasattr(w, 'to_obj') else {} for w in self.whens],
            'thens': [t.to_obj() if hasattr(t, 'to_obj') else {} for t in self.thens],
            'error': str(self.error) if self.error else None,
            'failed': self.failed,
            'features': self.features,
            'artifacts': self.artifacts
        }
    
    async def given_that(
        self,
        subject: Isubject,
        test_resource_configuration,
        artifactory: Callable[[str, Any], None],
        given_cb: Given,
        initial_values: Any,
        pm: Any
    ) -> Istore:
        raise NotImplementedError()
    
    async def after_each(
        self,
        store: Istore,
        key: str,
        artifactory: Callable[[str, Any], None],
        pm: Any
    ) -> Istore:
        return store
    
    def uber_catcher(self, e: Exception) -> None:
        self.error = e

    async def give(
        self,
        subject: Isubject,
        key: str,
        test_resource_configuration,
        tester: Callable[[Any], bool],
        artifactory: Callable[[str, Any], None],
        t_log: Callable[..., None],
        pm: Any,
        suite_ndx: int
    ) -> Istore:
        self.key = key
        t_log(f"\n {self.key}")
        t_log(f"\n Given: {self.name}")

        def given_artifactory(f_path: str, value: Any):
            artifactory(f"given-{key}/{f_path}", value)

        try:
            # Call given_that to set up the initial state
            self.store = await self.given_that(
                subject,
                test_resource_configuration,
                given_artifactory,
                self.given_cb,
                self.initial_values,
                pm
            )
        except Exception as e:
            self.failed = True
            self.error = e
            raise e

        try:
            # Process whens
            for when_ndx, when_step in enumerate(self.whens):
                await when_step.test(
                    self.store,
                    test_resource_configuration,
                    t_log,
                    pm,
                    f"suite-{suite_ndx}/given-{key}/when/{when_ndx}"
                )
            
            # Process thens
            for then_ndx, then_step in enumerate(self.thens):
                result = await then_step.test(
                    self.store,
                    test_resource_configuration,
                    t_log,
                    pm,
                    f"suite-{suite_ndx}/given-{key}/then-{then_ndx}"
                )
                tester(result)
        except Exception as e:
            self.error = e
            self.failed = True
            raise e
        finally:
            try:
                await self.after_each(self.store, self.key, given_artifactory, pm)
            except Exception as e:
                self.failed = True
                raise e

        return self.store
